lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -389,6 +500,182 @@ class DecomposeAtenReshapeOp : public OpRewritePattern<AtenReshapeOp> {
+                                          loc, inputTensors[i],
+                                          rewriter.create<Torch::ConstantIntOp>(
+                                              loc, rewriter.getI64IntegerAttr(j)))
+                                    : rewriter.create<Torch::ConstantIntOp>(
+                                          loc, rewriter.getI64IntegerAttr(inputShape[j])));
+      }
+      inputShapes.push_back(inputValueShape);
+    }
+
+    auto collectOperandDims = [resultTokens](
+                                  SmallVector<Value> operandShape,
+                                  SmallVector<char> operandTokens,
+                                  SmallVector<char> others,
+                                  SmallVectorImpl<int64_t> &contractingDims,
+                                  SmallVectorImpl<int64_t> &batchingDims,
+                                  SmallVector<char> &dotResultTokens,
+                                  SmallVector<Value> &dotResultShape) {
+      llvm::SmallDenseSet<char> othersSet(others.begin(), others.end());
+      llvm::SmallDenseSet<char> resultTokensSet(resultTokens.begin(),
+                                                resultTokens.end());
+      for (const auto &en : llvm::enumerate(operandTokens)) {
+        bool isResultToken = resultTokensSet.contains(en.value());
+        bool isOtherToken = othersSet.contains(en.value());
+        if (!isResultToken && isOtherToken) {
+          contractingDims.push_back(en.index());
+        } else if (isOtherToken) {
+          batchingDims.push_back(en.index());
+        } else {
+          dotResultTokens.push_back(en.value());
+          dotResultShape.push_back(operandShape[en.index()]);
+        }
+      }
+    };
+
+    Value constZero =
+        rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(0));
+    Value constOne =
+        rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(1));
+    Value constTwo =
+        rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(2));
+
+    while (inputTensors.size() > 1) {
+      SmallVector<int64_t> lhsContractingDims, lhsBatchingDims,
+          rhsContractingDims, rhsBatchingDims;
+      SmallVector<char> dotResultTokens;
+      SmallVector<Value> dotResultShape;
+      SmallVector<Value> lhsShape = inputShapes[0];
+      SmallVector<Value> rhsShape = inputShapes[1];
+      SmallVector<char> lhsTokens = inputTokens[0];
+      SmallVector<char> rhsTokens = inputTokens[1];
+      Value lhsTensor = inputTensors[0];
+      Value rhsTensor = inputTensors[1];
+      // Step 1: split input equation to input/result tokens
+      collectOperandDims(lhsShape, lhsTokens, rhsTokens, lhsContractingDims,
+                         lhsBatchingDims, dotResultTokens, dotResultShape);
+      collectOperandDims(rhsShape, rhsTokens, lhsTokens, rhsContractingDims,
+                         rhsBatchingDims, dotResultTokens, dotResultShape);
```

Comment:
I'm a bit confused by what is happening here. Can you explain this step to me? In particular, why are we collecting operands twice with `lhs` and `rhs` changing positions?

Reply:
> I'm a bit confused by what is happening here. Can you explain this step to me? In particular, why are we collecting operands twice with `lhs` and `rhs` changing positions?

In this function we are trying to find which dim should be contracted and which dim should be keep,  for example if we have equation "bcwd,bcdh->bcwh",  we could find b and c should be batched and d should be contracted